### PR TITLE
Server builder plugin

### PR DESCRIPTION
**Pull request by y-zeng**
Provide a scalable mechanism to add default services/functionality to servers without modifying include dependencies of existing files.

### CODE REVIEW

File: include/grpc++/impl/server_builder_plugin.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  // ServerBuilderOption::UpdatePlugins
+  virtual void ChangeArguments(const grpc::string& name, void* value) = 0;
+
+  virtual bool has_sync_methods() const { return false; }
+  virtual bool has_async_methods() const { return false; }
+};
+
+}  // namespace grpc
+
+#define DECLARE_PLUGIN(plugin_name)                                    \
```
Comment (ctiller) [inline include/grpc++/impl/server_builder_plugin.h]: macros need a GRPC_ prefix

Comment (y-zeng) [inline include/grpc++/impl/server_builder_plugin.h]: Fixed. Thanks for the review!

File: include/grpc++/server_builder.h [hunk 2]
```diff
@@ -113,6 +113,9 @@ class ServerBuilder {
   /// Return a running server which is ready for processing calls.
   std::unique_ptr<Server> BuildAndStart();
 
+  static void InternalAddPluginFactory(
```
Comment (yang-g) [inline include/grpc++/server_builder.h]: If we cannot move it to private, please add comment to ask users not to use it.

File: src/cpp/server/server_builder.cc [hunk 3]
```diff
@@ -41,9 +41,21 @@
 
 namespace grpc {
 
+static std::vector<std::unique_ptr<ServerBuilderPlugin> (*)()>* plugin_list;
```
Comment (yang-g) [inline src/cpp/server/server_builder.cc]: Name it g_plugin_factory_list?

File: test/cpp/end2end/server_builder_plugin_test.cc [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+  void UpdateArguments(ChannelArguments* arg) GRPC_OVERRIDE {}
+
+  void UpdatePlugins(
+      std::map<grpc::string, std::unique_ptr<ServerBuilderPlugin>>* plugins)
+      GRPC_OVERRIDE {
+    auto it = plugins->begin();
+    while (it != plugins->end()) {
+      plugins->erase(it++);
+    }
```
Comment (yang-g) [inline test/cpp/end2end/server_builder_plugin_test.cc]: plugins->clear()?

File: test/cpp/end2end/server_builder_plugin_test.cc [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+ private:
+  bool register_service_;
+};
+
+std::unique_ptr<ServerBuilderPlugin> CreateTestServerBuilderPlugin() {
+  return std::unique_ptr<ServerBuilderPlugin>(new TestServerBuilderPlugin());
+}
+
+void grpc_AddServerBuilderPlugin_reflection() {
```
Comment (yang-g) [inline test/cpp/end2end/server_builder_plugin_test.cc]: I think you want AddReflectionServerBuilderPlugin() as the name.

File: test/cpp/end2end/server_builder_plugin_test.cc [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+  static bool already_here = false;
+  if (already_here) return;
+  already_here = true;
+  ::grpc::ServerBuilder::InternalAddPluginFactory(
+      &CreateTestServerBuilderPlugin);
+}
+
+// Force AddServerBuilderPlugin() to be called at static initialization time.
+struct StaticPluginInitializer_reflection {
```
Comment (yang-g) [inline test/cpp/end2end/server_builder_plugin_test.cc]: ReflectionPluginInitializer ?

File: test/cpp/end2end/server_builder_plugin_test.cc [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+  StaticPluginInitializer_reflection() {
+    grpc_AddServerBuilderPlugin_reflection();
+  }
+} static_plugin_initializer_reflection_;
+
+class ServerBuilderPluginTest : public ::testing::TestWithParam<bool> {
```
Comment (yang-g) [inline test/cpp/end2end/server_builder_plugin_test.cc]: please document the param boolean

File: include/grpc++/server_builder.h (hunk 2 above)
Comment (y-zeng) [inline include/grpc++/server_builder.h]: Added comments. Thanks a lot for the review!

File: src/cpp/server/server_builder.cc (hunk 3 above)
Comment (y-zeng) [inline src/cpp/server/server_builder.cc]: Changed.

File: test/cpp/end2end/server_builder_plugin_test.cc (hunk 4 above)
Comment (y-zeng) [inline test/cpp/end2end/server_builder_plugin_test.cc]: Fixed.

File: test/cpp/end2end/server_builder_plugin_test.cc (hunk 5 above)
Comment (y-zeng) [inline test/cpp/end2end/server_builder_plugin_test.cc]: Fixed.

File: test/cpp/end2end/server_builder_plugin_test.cc (hunk 6 above)
Comment (y-zeng) [inline test/cpp/end2end/server_builder_plugin_test.cc]: Fixed.

File: test/cpp/end2end/server_builder_plugin_test.cc (hunk 7 above)
Comment (y-zeng) [inline test/cpp/end2end/server_builder_plugin_test.cc]: Added.

**Comment by yang-g** [comment]
LGTM
Please resolve conflicts.

**Pull request merged**